File: js/ColorGraph.js
"use strict";
var _=require('underscore');

// window.sandpile = window.sandpile || {};
var sandpile = {};

// ****************************************************************

sandpile.Graph = function(dim) {
	this.var = {};
	this.dim = dim // graph has fixed upper dim for matrix use
};

sandpile.Graph.prototype = {
	addVertex: function(vertex) {
		if (vertex<this.dim*this.dim){this.var[vertex] = [];}
	},
	addEdge: function(v1,v2) { // assumes v1,v2 in the object
		this.var[v1].push(v2);
		if (v1!==v2) {this.var[v2].push(v1);}
	},
	elimVertex: function(vertex) {
		delete this.var[vertex];
		for (var i in Object.keys(this.var)) {
			var elt = Object.keys(this.var)[i];
			var index = this.var[elt].indexOf(vertex);
			if (index!==-1) {this.var[elt].splice(index,1);}
		}
	},
	elimEdge: function(v1,v2) {
		if (this.hasEdge(v1,v2)) {
			var index = this.var[v1].indexOf(v2);
			this.var[v1].splice(index,1);
			index = this.var[v2].indexOf(v1);
			this.var[v2].splice(index,1);
		}		
	},
	hasEdge: function(v1,v2) {
		return this.var[v1].indexOf(v2)!==-1;
	},
	hasVertex: function(vertex) {
		return !_.isUndefined(this.var[vertex]);
	},
	listVertex: function() {
		return Object.keys(this.var);
	},
	listAdjacent: function(vertex) {
		return this.var[vertex];
	}
}
// ****************************************************************


sandpile.ColorGraph = function(dim) {
	this.graph = new sandpile.Graph(dim);
	this.colors = Array.apply(null, new Array(dim*dim)).map(Number.prototype.valueOf,0);
}

sandpile.ColorGraph.prototype = {
	addVertex: function(vertex) {
		this.graph.addVertex(vertex);
		this.colors[vertex] = 0
	},
	elimVertex: function(vertex) {
		this.graph.elimVertex(vertex);
		this.colors[vertex] = 0;
	},
	addEdge: function(v1,v2) {
		this.graph.addEdge(v1,v2);
	},
	elimEdge: function(v1,v2) {
		this.graph.elimEdge(v1,v2);
	},
	canColor: function(vertex,color) {
		var adjacents  = this.graph.listAdjacent(vertex);
		var can = true;
		for (var i  in adjacents) {
			can = can && (this.getColor(adjacents[i]) != color);
		}
		return can;
	},
	putColor: function(vertex,color) {
		if (this.canColor(vertex,color)) {
			this.colors[vertex] = color;
			return true;
		} return false;
	},
	forceColor: function(vertex,color) {
		this.colors[vertex] = color;
	},
	getColor: function(vertex) {
		return this.colors[vertex];
	},
	colorList: function() {
		return this.colors;
	},
	printMatrix: function() {
		var dim = this.graph.dim;
		for (var i = 0; i < dim; i++) {
			console.log(this.colors.slice(dim*i,dim*i+dim).join(' '))
		}
	}
}
var graph = new sandpile.ColorGraph(10);
graph.addVertex(3);
graph.addVertex(4);
graph.addVertex(2);
graph.addVertex(1);
graph.printMatrix()

File: java/ColorGraph.java
// Cole Ellison
// Color Graph Object

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.ArrayList;

class ColorGraph extends Graph {

	// CHANGE TO MATRIX	
	private HashMap<Integer,Integer> colors;

	public ColorGraph() {
		super();
		colors = new HashMap<Integer,Integer>();
	}

	public void eraseColor(int v) {
		colors.put(v,0);
	}

	public void eraseAll() {
		Iterator it = map.keySet().iterator();
		while (it.hasNext()) {
	    	colors.put((Integer)it.next(),0);
	    }
	}

	public void addVertex(int v) {
		super.addVertex(v);
		colors.put(v,0);
	}

	public void elimVertex(int v) {
		super.elimVertex(v);
		colors.remove(v);
	}

	public boolean canColor(int v, int color) {
		HashSet<Integer> adjacents = super.listAdjacent(v);
		boolean can = true;
		for (Integer temp : adjacents) {
			can = can && (getColor(temp) != color);
		}
	    return can;
	}

	public boolean putColor(int v, int color) {
		if (canColor(v,color)==true) {
			colors.put(v,color);
			return true;
		} else {return false;}
	}

	public void forceColor(int v, int color) {
		colors.put(v,color);
	}

	public int getColor(int v) {
		return colors.get(v);
	}

	public ArrayList colorList() {
		ArrayList<Integer> c = new ArrayList<Integer>();
		for (Integer vertex : colors.keySet()) {
			c.add(getColor(vertex));
		}
		return c;
	}

	public boolean hasColor(int v) {
		return getColor(v)!=0;
	}

	public ArrayList<Integer> possibleColors(int v,int max) {
		if (hasColor(v)!=true) { // solo busca vertices sin color
			HashSet<Integer> adjacent = listAdjacent(v);
			ArrayList<Integer> nums = new ArrayList<Integer>();
			for (int i=1; i<max+1; i++) { // busca color no contenido en los adjacent
				nums.add(i);
			}
			for (Integer temp : adjacent) {
				int index = nums.indexOf(getColor(temp));
				if (index != -1) {
					nums.remove(index);
				}
			}
			return nums;
		} else {
			return new ArrayList<Integer>();
		}
	}

	public static void main(String[] args) {
		ColorGraph graph = new ColorGraph();
		graph.addVertex(3);
		graph.addVertex(4);
		graph.addVertex(2);
		graph.addVertex(1);
		System.out.println(graph.listVertex());
		graph.addEdge(3,4);
		graph.addEdge(4,1);
		graph.addEdge(3,3);
		System.out.println(graph.listAdjacent(3));
		graph.putColor(3,10);
		System.out.println(graph.colorList());
		System.out.println(graph.canColor(4,10));
		graph.elimEdge(3,3);
		System.out.println(graph.listAdjacent(3));
		graph.elimVertex(3);
		System.out.println(graph.listAdjacent(4));
		graph.elimVertex(1);
		System.out.println(graph.listVertex());
		System.out.println(graph.listAdjacent(4));
	}
}